src/Raven.Server/Documents/Indexes/Static/Roslyn/Rewriters/DynamicLambdaExpressionsRewriter.cs

```diff
@@ -209,6 +222,39 @@ private SyntaxNode ModifyLambdaForSelectMany(LambdaExpressionSyntax node, Invoca
             return node;
         }
 
+        private static bool IsSelectManyCalledOnValueCollection(InvocationExpressionSyntax invocation)
+        {
+            if (invocation.Expression is MemberAccessExpressionSyntax e1)
+            {
+                if (e1.Expression is ObjectCreationExpressionSyntax e2)
+                {
+                    if (e2.ArgumentList?.Arguments[0].Expression is MemberAccessExpressionSyntax m1)
+                    {
+                        if (m1.Name.Identifier.Text == "Values" || m1.Name.Identifier.Text == "Keys")
+                            return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        private static bool DoesLambdaReturnDictionary(LambdaExpressionSyntax node)
```

Comment:
Maybe `IsDictionaryReturned()`?

Reply:
Renamed, additionally I renamed variables in `IsSelectManyCalledOnValueCollection` to be more meaningful and added an additional check